File: Medium/279. Perfect Squares/solution (4).cpp
// 279. Perfect Squares

// Runtime: 267 ms, faster than 57.02% of C++ online submissions for Perfect Squares.

// Memory Usage: 33.7 MB, less than 14.65% of C++ online submissions for Perfect Squares.


class Solution {
public:
    // Breadth-First Search
    int numSquares(int n) {
        const auto squareNums {getSquareNums(n)};
        set<int> nodes;
        nodes.insert(n);

        int count {1};
        while (!nodes.empty()) {
            set<int> nextNodes;
            for (const auto remain : nodes) {
                for (const auto square : squareNums) {
                    if (remain == square) {
                        return count;
                    } else if (remain >= square) {
                        nextNodes.insert(remain - square);
                    }
                }
            }

            ++count;
            nodes = nextNodes;
        }

        return count;
    }

private:
    static vector<int> getSquareNums(const int n) {
        vector<int> nums;
        for (int i {1}; i < (int)pow(n, 0.5) + 1; ++i) {
            nums.push_back(i * i);
        }

        return nums;
    }
};

File: Easy/383. Ransom Note/solution (4).py
# 383. Ransom Note

# Runtime: 160 ms, faster than 5.37% of Python3 online submissions for Ransom Note.

# Memory Usage: 14.6 MB, less than 11.24% of Python3 online submissions for Ransom Note.


class Solution:
    # Sorting and Stacks
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        if len(ransomNote) > len(magazine):
            return False

        ransomNote = sorted(ransomNote, reverse=True)
        magazine = sorted(magazine, reverse=True)
        while len(ransomNote) > 0 and len(magazine) > 0:
            if ransomNote[-1] == magazine[-1]:
                ransomNote.pop()
                magazine.pop()
            elif ransomNote[-1] > magazine[-1]:
                magazine.pop()
            else:
                return False
        return len(ransomNote) == 0